File: src/Envirotrax.App/Envirotrax.App.Server/Data/Models/Backflow/BackflowTest.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using Envirotrax.App.Server.Data.Models.Professionals;
using Envirotrax.App.Server.Data.Models.Sites;
using Envirotrax.App.Server.Data.Models.States;
using Envirotrax.App.Server.Data.Models.Users;
using Envirotrax.App.Server.Data.Models.WaterSuppliers;
using Envirotrax.Common.Data.Attributes;
using Envirotrax.Common.Data.Models;
using Microsoft.EntityFrameworkCore;
using Microsoft.EntityFrameworkCore.Metadata.Builders;

namespace Envirotrax.App.Server.Data.Models.Backflow;

[Table("BackflowTests")]
public class BackflowTest : TenantModel<WaterSupplier>, IAuditableModel<AppUser>
{
    [AppPrimaryKey(true)]
    public int Id { get; set; }

    public int? SiteId { get; set; }
    public Site? Site { get; set; }

    [StringLength(50)]
    public string? SubmissionId { get; set; }

    [StringLength(50)]
    public string? JobNumber { get; set; }

    // BPAT (Backflow Prevention Assembly Tester)
    public int? ProfessionalId { get; set; }
    public int? BpatId { get; set; }
    public ProfessionalUser? Bpat { get; set; }

    [StringLength(50)]
    public string? BpatLicenseNumber { get; set; }

    public DateTime? BpatLicenseExpiration { get; set; }

    [StringLength(100)]
    public string? BpatCompanyName { get; set; }

    [StringLength(100)]
    public string? BpatContactName { get; set; }

    [StringLength(200)]
    public string? BpatAddress { get; set; }

    [StringLength(50)]
    public string? BpatCity { get; set; }

    public int? BpatStateId { get; set; }
    public State? BpatState { get; set; }

    [StringLength(20)]
    public string? BpatZip { get; set; }

    [StringLength(50)]
    public string? BpatWorkNumber { get; set; }

    [StringLength(50)]
    public string? BpatCellNumber { get; set; }

    // Property / Site fields
    [StringLength(50)]
    public string? AccountNumber { get; set; }

    [StringLength(100)]
    public string? PropertyBusinessName { get; set; }

    public int PropertyType { get; set; }

    [StringLength(50)]
    public string? PropertyStreetNumber { get; set; }

    [StringLength(100)]
    public string? PropertyStreetName { get; set; }

    [StringLength(50)]
    public string? PropertyNumber { get; set; }

    [StringLength(50)]
    public string? PropertyCity { get; set; }

    public int? PropertyStateId { get; set; }
    public State? PropertyState { get; set; }

    [StringLength(20)]
    public string? PropertyZip { get; set; }

    // Mailing address
    [StringLength(100)]
    public string? MailingCompanyName { get; set; }

    [StringLength(100)]
    public string? MailingContactName { get; set; }

    [StringLength(50)]
    public string? MailingStreetNumber { get; set; }

    [StringLength(100)]
    public string? MailingStreetName { get; set; }

    [StringLength(50)]
    public string? MailingNumber { get; set; }

    [StringLength(50)]
    public string? MailingCity { get; set; }

    public int? MailingStateId { get; set; }
    public State? MailingState { get; set; }

    [StringLength(20)]
    public string? MailingZip { get; set; }

    [StringLength(50)]
    public string? MailingPhoneNumber { get; set; }

    [StringLength(100)]
    public string? MailingEmailAddress { get; set; }

    // Device info
    [StringLength(50)]
    public string? DeviceType { get; set; }

    [StringLength(100)]
    public string? Manufacturer { get; set; }

    [StringLength(100)]
    public string? Model { get; set; }

    [StringLength(50)]
    public string? Size { get; set; }

    [StringLength(100)]
    public string? SerialNumber { get; set; }

    public bool UnknownSerialNumber { get; set; }

    [StringLength(100)]
    public string? Manufacturer2 { get; set; }

    [StringLength(100)]
    public string? Model2 { get; set; }

    [StringLength(50)]
    public string? Size2 { get; set; }

    [StringLength(100)]
    public string? SerialNumber2 { get; set; }

    [StringLength(200)]
    public string? LocationDescription { get; set; }

    [StringLength(100)]
    public string? HazardType { get; set; }

    [StringLength(200)]
    public string? HazardTypeOtherDescription { get; set; }

    // Test info
    public BackflowReasonForTest ReasonForTest { get; set; }

    [StringLength(100)]
    public string? ReplacementAssembly { get; set; }

    public DateTime? InstallationDate { get; set; }
    public DateTime? TestDate { get; set; }
    public DateTime? InitialTestDate { get; set; }
    public DateTime? RepairTestDate { get; set; }
    public DateTime? FinalTestDate { get; set; }
    public DateTime? ExpirationDate { get; set; }

    public BackflowTestResult TestResult { get; set; }

    public bool ProperlyInstalled { get; set; }
    public bool NonPotable { get; set; }

    // Gauge
    [StringLength(100)]
    public string? GaugeManufacturer { get; set; }

    [StringLength(100)]
    public string? GaugeModel { get; set; }

    [StringLength(50)]
    public string? GaugeSerialNumber { get; set; }

    public DateTime? GaugeLastCalibrationDate { get; set; }
    public bool GaugeNonPotable { get; set; }

    // Optional fields
    [StringLength(50)]
    public string? MeterNumber { get; set; }

    [Precision(5, 2)]
    public decimal? MeterReadingBefore { get; set; }
    public bool MeterRegisters { get; set; }
    [Precision(5, 2)]
    public decimal? MeterReadingAfter { get; set; }

    [StringLength(50)]
    public string? PermitNumber { get; set; }

    public bool Ossf { get; set; }

    // Status
    public string? Comments { get; set; }
    public bool IsCurrent { get; set; }
    public bool Disapproved { get; set; }
    public bool OutOfService { get; set; }
    public DateTime? OutOfServiceDate { get; set; }

    // Approval / Payment
    public DateTime? ApprovalDate { get; set; }

    public int? ApprovedById { get; set; }
    public WaterSupplierUser? ApprovedBy { get; set; }

    [StringLength(100)]
    public string? TransactionId { get; set; }

    public DateTime? TransactionDate { get; set; }

    [Precision(19, 4)]
    public decimal Amount { get; set; }

    [Precision(19, 4)]
    public decimal AmountShare { get; set; }

    public bool EmailPdf { get; set; }

    // Rejection
    public bool Rejected { get; set; }

    public int? RejectedById { get; set; }
    public WaterSupplierUser? RejectedBy { get; set; }

    public DateTime? RejectedDate { get; set; }

    public string? RejectedReason { get; set; }

    // Validation flags
    public bool NeedsValidation { get; set; }
    public bool ValidationNewSite { get; set; }
    public bool ValidationSiteInformationChanged { get; set; }
    public bool ValidationUnknownSerialNumber { get; set; }
    public bool ValidationDeviceInformationChanged { get; set; }

    // Initial test readings - main assembly
    [Precision(5, 2)]
    public decimal? InitCV1HeldPSID { get; set; }
    public bool InitCV1ClosedTight { get; set; }
    public bool InitCV1Leaked { get; set; }

    [Precision(5, 2)]
    public decimal? InitCV2HeldPSID { get; set; }
    public bool InitCV2ClosedTight { get; set; }
    public bool InitCV2Leaked { get; set; }

    [Precision(5, 2)]
    public decimal? InitRVOpenedPSID { get; set; }
    public bool InitRVDidNotOpen { get; set; }

    [Precision(5, 2)]
    public decimal? InitBCHeldPSID { get; set; }
    public bool InitBCClosedTight { get; set; }
    public bool InitBCLeaked { get; set; }

    [Precision(5, 2)]
    public decimal? InitPvbAirInletOpenedPSID { get; set; }
    public bool InitPvbAirInletDidNotOpen { get; set; }
    public bool InitPvbAirInletFullyOpened { get; set; }

    [Precision(5, 2)]
    public decimal? InitPvbCVHeldPSID { get; set; }
    public bool InitPvbCVLeaked { get; set; }

    // Air gap
    public bool AirGapValid { get; set; }

    // Repairs (stored as comma-separated text, e.g. "Cleaned, Replaced Disc, Replaced Spring")
    [StringLength(200)]
    public string? RepairCV1 { get; set; }

    [StringLength(100)]
    public string? RepairCV1Details { get; set; }

    [StringLength(200)]
    public string? RepairCV2 { get; set; }

    [StringLength(100)]
    public string? RepairCV2Details { get; set; }

    [StringLength(200)]
    public string? RepairRV { get; set; }

    [StringLength(100)]
    public string? RepairRVDetails { get; set; }

    [StringLength(200)]
    public string? RepairBC { get; set; }

    [StringLength(100)]
    public string? RepairBCDetails { get; set; }

    // Final test readings - main assembly
    [Precision(5, 2)]
    public decimal? FinalCV1HeldPSID { get; set; }
    public bool FinalCV1ClosedTight { get; set; }

    [Precision(5, 2)]
    public decimal? FinalCV2HeldPSID { get; set; }
    public bool FinalCV2ClosedTight { get; set; }

    [Precision(5, 2)]
    public decimal? FinalRVOpenedPSID { get; set; }

    [Precision(5, 2)]
    public decimal? FinalBCHeldPSID { get; set; }
    public bool FinalBCClosedTight { get; set; }

    [Precision(5, 2)]
    public decimal? FinalPvbAirInletOpenedPSID { get; set; }
    public bool FinalPvbAirInletFullyOpened { get; set; }

    [Precision(5, 2)]
    public decimal? FinalPvbCVHeldPSID { get; set; }

    // Bypass assembly readings (DCD/RPPD device types only)
    [Precision(5, 2)]
    public decimal? InitCV1HeldPSID2 { get; set; }
    public bool InitCV1ClosedTight2 { get; set; }
    public bool InitCV1Leaked2 { get; set; }

    [Precision(5, 2)]
    public decimal? InitCV2HeldPSID2 { get; set; }
    public bool InitCV2ClosedTight2 { get; set; }
    public bool InitCV2Leaked2 { get; set; }

    [Precision(5, 2)]
    public decimal? InitRVOpenedPSID2 { get; set; }
    public bool InitRVDidNotOpen2 { get; set; }

    [StringLength(200)]
    public string? RepairCV12 { get; set; }

    [StringLength(100)]
    public string? RepairCV1Details2 { get; set; }

    [StringLength(200)]
    public string? RepairCV22 { get; set; }

    [StringLength(100)]
    public string? RepairCV2Details2 { get; set; }

    [StringLength(200)]
    public string? RepairRV2 { get; set; }

    [StringLength(100)]
    public string? RepairRVDetails2 { get; set; }

    [Precision(5, 2)]
    public decimal? FinalCV1HeldPSID2 { get; set; }
    public bool FinalCV1ClosedTight2 { get; set; }
    public bool FinalCV2ClosedTight2 { get; set; }

    [Precision(5, 2)]
    public decimal? FinalRVOpenedPSID2 { get; set; }

    // Audit
    public int? CreatedById { get; set; }
    public AppUser? CreatedBy { get; set; }
    public DateTime CreatedTime { get; set; }
    public int? UpdatedById { get; set; }
    public AppUser? UpdatedBy { get; set; }
    public DateTime? UpdatedTime { get; set; }
    public int? DeletedById { get; set; }
    public AppUser? DeletedBy { get; set; }
    public DateTime? DeletedTime { get; set; }
}

public class BackflowTestConfiguration : IEntityTypeConfiguration<BackflowTest>
{
    public void Configure(EntityTypeBuilder<BackflowTest> builder)
    {
        builder.HasOne<ProfessionalUser>(bt => bt.Bpat)
            .WithMany()
            .HasForeignKey(bt => new { bt.ProfessionalId, bt.BpatId })
            .HasPrincipalKey(pu => new { pu.ProfessionalId, pu.UserId });
    }
}

File: src/Envirotrax.Admin/envirotrax.admin.client/src/app/shared/models/backflow/backflow-test.ts
import { State } from '../lookup/state';
import { PropertyType } from '../sites/site';
import { WaterSupplier } from '../water-suppliers/water-supplier';

export enum BackflowPaymentStatus {
    Paid = 1,
    Unpaid = 2
}

export enum BackflowTestResult {
    Pass = 0,
    Fail = 1,
    PassAfterRepairs = 2
}

export enum BackflowReasonForTest {
    AnnualTest = 0,
    NewInstallation = 1,
    ExistingInstallation = 2,
    Replacement = 3,
    Repair = 4,
    AnnualTestAfterRepairs = 5
}

export enum BackflowDeviceType {
    DC = 'DC',
    DCD = 'DCD',
    DCD2 = 'DCD2',
    RP = 'RP',
    RPPD = 'RPPD',
    RPPD2 = 'RPPD2',
    PVB = 'PVB',
    SVB = 'SVB',
    AG = 'AG'
}

export const backflowDeviceTypeDescriptions: Record<string, string> = {
    [BackflowDeviceType.DC]: 'Double Check Valve',
    [BackflowDeviceType.DCD]: 'Double Check Detector',
    [BackflowDeviceType.DCD2]: 'Double Check Detector Type II',
    [BackflowDeviceType.RP]: 'Reduced Pressure Principle',
    [BackflowDeviceType.RPPD]: 'Reduced Pressure Principle Detector',
    [BackflowDeviceType.RPPD2]: 'Reduced Pressure Principle Detector Type II',
    [BackflowDeviceType.PVB]: 'Pressure Vacuum Breaker',
    [BackflowDeviceType.SVB]: 'Spill-Resistant Pressure Vacuum Breaker',
    [BackflowDeviceType.AG]: 'Air Gap'
};

export class BackflowTest {
    id?: number;
    waterSupplier?: WaterSupplier;
    accountNumber?: string;
    createdTime?: string;
    testDate?: string;
    expirationDate?: string;
    testResult?: BackflowTestResult;
    isCurrent?: boolean;
    outOfService?: boolean;
    disapproved?: boolean;
    rejected?: boolean;
    transactionId?: string;
    serialNumber?: string;
    serialNumber2?: string;
    manufacturer?: string;
    model?: string;
    size?: string;
    deviceType?: string;
    hazardType?: string;
    locationDescription?: string;
    propertyType?: PropertyType;
    propertyBusinessName?: string;
    propertyStreetNumber?: string;
    propertyStreetName?: string;
    propertyNumber?: string;
    propertyCity?: string;
    propertyState?: State;
    propertyZip?: string;
    bpatCompanyName?: string;
    bpatContactName?: string;
    bpatLicenseNumber?: string;
}

export class BackflowTestReviewer {
    id?: number;
    contactName?: string;
    emailAddress?: string;
}

export class BackflowTestSite {
    id?: number;
    accountNumber?: string;
}

export class BackflowTestBpat {
    id?: number;
    emailAddress?: string;
    contactName?: string;
}

export class BackflowTestCounts {
    recordLogCount?: number;
}

export class BackflowTestDetails {
    id?: number;
    waterSupplier?: WaterSupplier;
    site?: BackflowTestSite;
    bpat?: BackflowTestBpat;
    submissionId?: string;
    accountNumber?: string;

    isCurrent?: boolean;
    outOfService?: boolean;
    outOfServiceDate?: string;
    disapproved?: boolean;
    rejected?: boolean;
    needsValidation?: boolean;
    forceRenewal?: boolean;
    forceRenewalYears?: number;
    backflowScheduleMonth?: number;
    renewalRequired?: boolean;
    needsRenewalCheck?: boolean;

    approvalDate?: string;
    approvedBy?: BackflowTestReviewer;
    rejectedDate?: string;
    rejectedBy?: BackflowTestReviewer;
    rejectedReason?: string;

    validationNewSite?: boolean;
    validationSiteInformationChanged?: boolean;
    validationUnknownSerialNumber?: boolean;
    validationDeviceInformationChanged?: boolean;
    validationNotes?: string;

    testDate?: string;
    expirationDate?: string;
    installationDate?: string;

    transactionId?: string;
    transactionDate?: string;
    amount?: number;
    amountShare?: number;

    bpatCompanyName?: string;
    bpatJobTitle?: string;
    bpatContactName?: string;
    bpatAddress?: string;
    bpatCity?: string;
    bpatState?: State;
    bpatZip?: string;
    bpatWorkNumber?: string;
    bpatCellNumber?: string;
    bpatLicenseNumber?: string;
    bpatLicenseExpiration?: string;

    propertyType?: PropertyType;
    propertyBusinessName?: string;
    propertyStreetNumber?: string;
    propertyStreetName?: string;
    propertyNumber?: string;
    propertyCity?: string;
    propertyState?: State;
    propertyZip?: string;

    mailingCompanyName?: string;
    mailingContactName?: string;
    mailingStreetNumber?: string;
    mailingStreetName?: string;
    mailingNumber?: string;
    mailingCity?: string;
    mailingState?: State;
    mailingZip?: string;

    deviceType?: string;
    manufacturer?: string;
    model?: string;
    size?: string;
    serialNumber?: string;
    unknownSerialNumber?: boolean;
    manufacturer2?: string;
    model2?: string;
    size2?: string;
    serialNumber2?: string;
    locationDescription?: string;
    hazardType?: string;
    hazardTypeOtherDescription?: string;

    testResult?: BackflowTestResult;
    jobNumber?: string;
    reasonForTest?: BackflowReasonForTest;
    replacementAssembly?: string;
    properlyInstalled?: boolean;
    nonPotable?: boolean;

    gaugeManufacturer?: string;
    gaugeModel?: string;
    gaugeSerialNumber?: string;
    gaugeLastCalibrationDate?: string;
    gaugeNonPotable?: boolean;

    initialTestDate?: string;

    initCV1HeldPSID?: number;
    initCV1ClosedTight?: boolean;
    initCV1Leaked?: boolean;
    initCV2HeldPSID?: number;
    initCV2ClosedTight?: boolean;
    initCV2Leaked?: boolean;
    initRVOpenedPSID?: number;
    initRVDidNotOpen?: boolean;
    initBCHeldPSID?: number;
    initBCClosedTight?: boolean;
    initBCLeaked?: boolean;
    initPvbAirInletOpenedPSID?: number;
    initPvbAirInletDidNotOpen?: boolean;
    initPvbAirInletFullyOpened?: boolean;
    initPvbCVHeldPSID?: number;
    initPvbCVLeaked?: boolean;

    initCV1HeldPSID2?: number;
    initCV1ClosedTight2?: boolean;
    initCV1Leaked2?: boolean;
    initCV2HeldPSID2?: number;
    initCV2ClosedTight2?: boolean;
    initCV2Leaked2?: boolean;
    initRVOpenedPSID2?: number;
    initRVDidNotOpen2?: boolean;

    repairCV1?: string;
    repairCV2?: string;
    repairRV?: string;
    repairBC?: string;
    repairCV12?: string;
    repairCV22?: string;
    repairRV2?: string;
    repairPvbAirInlet?: string;
    repairPvbCV?: string;

    repairCV1Details?: string;
    repairCV2Details?: string;
    repairRVDetails?: string;
    repairBCDetails?: string;
    repairCV1Details2?: string;
    repairCV2Details2?: string;
    repairRVDetails2?: string;
    repairPvbAirInletDetails?: string;
    repairPvbCVDetails?: string;

    repairTestDate?: string;

    finalCV1HeldPSID?: number;
    finalCV1ClosedTight?: boolean;
    finalCV2HeldPSID?: number;
    finalCV2ClosedTight?: boolean;
    finalRVOpenedPSID?: number;
    finalBCHeldPSID?: number;
    finalBCClosedTight?: boolean;
    finalPvbAirInletOpenedPSID?: number;
    finalPvbAirInletFullyOpened?: boolean;
    finalPvbCVHeldPSID?: number;

    finalCV1HeldPSID2?: number;
    finalCV1ClosedTight2?: boolean;
    finalCV2HeldPSID2?: number;
    finalCV2ClosedTight2?: boolean;
    finalRVOpenedPSID2?: number;

    meterNumber?: string;
    meterRegisters?: boolean;
    meterReadingBefore?: number;
    meterReadingAfter?: number;

    airGapValid?: boolean;

    ossf?: boolean;
    rainFreezeSensorInstalled?: boolean;
    rainFreezeSensorWorkingProperly?: boolean;
    permitNumber?: string;

    comments?: string;

    assemblyImageUrl?: string;
    serialNumberImageUrl?: string;
    bypassAssemblyImageUrl?: string;
    bypassSerialNumberImageUrl?: string;
    airGapImageUrl?: string;

    showRainSensor?: boolean;
    showOSSF?: boolean;
    showPermitNumber?: boolean;
}
